common/prque/lazyqueue.go

```diff
@@ -0,0 +1,181 @@
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+package prque
+
+import (
+	"container/heap"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+)
+
+// LazyQueue is a priority queue data structure where priorities can change over
+// time and are only evaluated on demand.
+// Two callbacks are required:
+// - priority evaluates the actual priority of an item
+// - maxPriority gives an upper estimate for the priority in any moment between
+//   now and the given absolute time
+// If the upper estimate is exceeded then Update should be called for that item.
+// A global Refresh function should also be called periodically.
+type LazyQueue struct {
+	clock mclock.Clock
+	// Items are stored in one of two internal queues ordered by estimated max
+	// priority until the next and the next-after-next refresh. Update and Refresh
+	// always places items in queue[1].
+	queue       [2]*sstack
+	popQueue    *sstack
+	period      time.Duration
+	maxUntil    mclock.AbsTime
+	indexOffset int
+	setIndex    SetIndexCallback
+	priority    PriorityCallback
+	maxPriority MaxPriorityCallback
+}
+
+type (
+	PriorityCallback    func(data interface{}, now mclock.AbsTime) int64   // actual priority callback
+	MaxPriorityCallback func(data interface{}, until mclock.AbsTime) int64 // estimated maximum priority callback
+)
+
+// NewLazyQueue creates a new lazy queue
+func NewLazyQueue(setIndex SetIndexCallback, priority PriorityCallback, maxPriority MaxPriorityCallback, clock mclock.Clock, refreshPeriod time.Duration) *LazyQueue {
+	q := &LazyQueue{
+		popQueue:    newSstack(nil),
+		setIndex:    setIndex,
+		priority:    priority,
+		maxPriority: maxPriority,
+		clock:       clock,
+		period:      refreshPeriod}
+	q.Reset()
+	q.Refresh()
+	return q
+}
+
+// Reset clears the contents of the queue
+func (q *LazyQueue) Reset() {
+	q.queue[0] = newSstack(q.setIndex0)
+	q.queue[1] = newSstack(q.setIndex1)
+}
+
+// Refresh should be called at least with the frequency specified by the refreshPeriod parameter
+func (q *LazyQueue) Refresh() {
+	q.maxUntil = q.clock.Now() + mclock.AbsTime(q.period)
+	for q.queue[0].Len() != 0 {
+		q.Push(heap.Pop(q.queue[0]).(*item).value)
+	}
+	q.queue[0], q.queue[1] = q.queue[1], q.queue[0]
+	q.indexOffset = 1 - q.indexOffset
+	q.maxUntil += mclock.AbsTime(q.period)
```

Comment:
The idea is that queue[0] is ordered by max priority at next Resfresh, queue[1] at next-after-next refresh. Update always puts items into queue[1] so maxUntil should hold the expected time of the next-after-next refresh.

Reply:
If so, when we pop some items, the priority of Q0 items are based on `next`, while the priority of Q1 items is based on `next-after-next`. How can they be compared?

Reply:
All we care about here is that estimated priority is at least as high as the actual priority. This is equally true for Q0 and Q1 items. The only difference is that in order to keep this property true at all times Q0 items need to be updated in the next refresh cycle if they don't get updated sooner for some other reason.

Reply:
Why don't we insert all newly added items to the same Q? 
For example, now we shuffle all items from Q1 -> Q0, so all item's priority gets updated. 
Before the next refresh, all new items are still inserted to Q0, so that **all items can have same base time for estimated priority**.

Reply:
Here is a hack of a simplified lazy-queue which only uses another queue as a cache.

Reply:
I think I understand your version now, it took a bit of time :) So what you are doing here is that you always keep the actual items in queue[0] (and keep them there even if a manual update happens) and then move all items from one queue to the other and swap them. Also, you call the maxPriority callback with the end of the current refresh period as a parameter. This might work but it might also be inefficient. Once my version does an externally initiated Update it tries to prevent updating it many times by estimating its max priority for the next cycle. So if the estimate increases frequently but only a little bit each time (which is exactly what is happening in our use case with each processed request) then it prevents frequent queue operations. Putting it in queue[1] is also an efficiency improvement because once we have updated it we know that it does not need to be updated in the end of the current cycle because its priority was already estimated for the next one. So what happens in my version is that every item is usually moved only once per refresh cycle, either during an Update or the Refresh itself.

Reply:
Thanks for the explanation, I get your idea now. The accurate of estimated priority is not matter at all. We just use estimated priority to pick some items that are most likely to have the highest actual priority and then use a popQueue to pick the item with the highest actual priority one by one. And your approach is efficient.